### PR TITLE
Handle default: keyword within switch statement

### PR DESCRIPTION
**Pull request by mastef**
This should take care of #273 : Where `default:` is being used in a switch statement before other cases.

- It will move the cases to the `default` value instead
- It will add a comment with `/* covers case "d", "e": */`
- It will display the cases' comments before the `default:` statement
- It should handle the `fall-through` properly if the `default:` is not in the end of the `switch`

Regarding unit tests, I was testing with the code below : 

```as3
package {
    public class SwitchTest {
        public function SwitchTest() {

            var param1 = "g";
            switch(param1)
            {
                // default should cover A
                default:
                // comment before A
                case "a":
                   this.val = "a";
                   break;
                case "b":
                   this.val = "b";
                   break;
                case "c":
                   this.val = "c";
                   break;
                case "d":
                   this.val = "d";
                   break;
                // comment before E
                case "e":
                   this.val = "e";
                   break;
                // comment before F
                case "f":
                   this.val = "f";
            }

            var param2 = "a";
            switch(param2)
            {
                // comment before A
                case "a":
                   this.val = "a";
                   break;
                case "b":
                   this.val = "b";
                   break;
                // default should cover C
                // default should happen after B and before D
                default:
                case "c":
                   this.val = "c";
                   break;
                case "d":
                   this.val = "d";
                   break;
                // comment before E
                case "e":
                   this.val = "e";
                   break;
                // comment before F
                case "f":
                   this.val = "f";
            }

            var test1a = "j";
            // default should include c
            // default should return 3
            // default should be last
            switch(test1a)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                default:
                case "c":
                    return 3;
            }

            var test1b = "b";
            // default should include c
            // default should return 3
            // default should be last
            switch(test1b)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                case "c":
                default:
                    return 3;
            }

            var test1c = "j";
            // default could include a, b, c
            // default should return 1
            // default should be last
            switch(test1c)
            {
                // comment before case a
                case "a":
                    return 1;
                // comment before case b
                case "b":
                    return 1;
                // comment before default
                default:
                // comment before case c
                case "c":
                    return 1;
            }

            // default should return 4
            // default should come before d
            var test2 = "d";
            switch(test2)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                case "c":
                    return 3;
                default:
                    return 4;
                case "e":
                    return 5;
            }

            // d and e should be the same
            // no default
            var test3 = "d";
            switch(test3)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                case "c":
                    return 3;
                case "d":
                case "e":
                    return 5;
                case "f":
                    return 6;
            }

            // default should include d
            // default should be empty
            // default should be before e
            var test4 = "j";
            switch(test4)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                case "c":
                    return 3;
                default:
                case "d":
                    break;
                case "e":
                    return 5;
                case "f":
                    return 6;
            }

            // default should include d and e
            // default should return 4
            // default should be before f
            var test5 = "d";
            switch(test5)
            {
                case "a":
                    return 1;
                case "b":
                    return 2;
                case "c":
                    return 3;
                default:
                case "d":
                case "e":
                    return 5;
                case "f":
                    return 6;
            }

            // default should include d, e and f
            // default should return 6
            // default should be in the end
            var test6 = "d";
            switch(test6)
            {
                case "a":
                case "b":
                case "c":
                    return 3;
                case "d":
                case "e":
                case "f":
                default:
                    return 6;
            }

            // test with comments before every case
            var test7 = "d";
            switch(test7)
            {
                // comment before a
                case "a":
                    return 1;
                // comment before b
                case "b":
                    return 2;
                // comment before c
                case "c":
                    return 3;
                // comment before default
                default:
                // comment before d
                case "d":
                // comment before e
                case "e":
                    return 5;
                // comment before f
                case "f":
                    return 6;
            }
        }
    }
}
```

Which produced this output : 

```haxe
class SwitchTest
{
    public function new()
    {
        var param1 = "g";
        switch (param1)
        {
            /* covers case "a": */
            // default should cover A
            // comment before A
            default:
                this.val = "a";
            case "b":
                this.val = "b";
            case "c":
                this.val = "c";
            case "d":
                this.val = "d";
            // comment before E
            case "e":
                this.val = "e";
            // comment before F
            case "f":
                this.val = "f";
        }
        
        var param2 = "a";
        switch (param2)
        {
            // comment before A
            case "a":
                this.val = "a";
            case "b":
                this.val = "b";
            /* covers case "c": */
            // default should cover C
            // default should happen after B and before D
            default:
                this.val = "c";
            case "d":
                this.val = "d";
            // comment before E
            case "e":
                this.val = "e";
            // comment before F
            case "f":
                this.val = "f";
        }
        
        var test1a = "j";
        // default should include c
        // default should return 3
        // default should be last
        switch (test1a)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            /* covers case "c": */
            default:
                return 3;
        }
        
        var test1b = "b";
        // default should include c
        // default should return 3
        // default should be last
        switch (test1b)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            case "c":
                return 3;
            default:
                return 3;
        }
        
        var test1c = "j";
        // default could include a, b, c
        // default should return 1
        // default should be last
        switch (test1c)
        {
            // comment before case a
            case "a":
                return 1;
            // comment before case b
            case "b":
                return 1;
            /* covers case "c": */
            // comment before default
            // comment before case c
            default:
                return 1;
        }
        
        // default should return 4
        // default should come before d
        var test2 = "d";
        switch (test2)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            case "c":
                return 3;
            default:
                return 4;
            case "e":
                return 5;
        }
        
        // d and e should be the same
        // no default
        var test3 = "d";
        switch (test3)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            case "c":
                return 3;
            case "d", "e":
                return 5;
            case "f":
                return 6;
        }
        
        // default should include d
        // default should be empty
        // default should be before e
        var test4 = "j";
        switch (test4)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            case "c":
                return 3;
            /* covers case "d": */
            default:
            case "e":
                return 5;
            case "f":
                return 6;
        }
        
        // default should include d and e
        // default should return 4
        // default should be before f
        var test5 = "d";
        switch (test5)
        {
            case "a":
                return 1;
            case "b":
                return 2;
            case "c":
                return 3;
            /* covers case "d", "e": */
            default:
                return 5;
            case "f":
                return 6;
        }
        
        // default should include d, e and f
        // default should return 6
        // default should be in the end
        var test6 = "d";
        switch (test6)
        {
            case "a", "b", "c":
                return 3;
            case "d", "e", "f":
                return 6;
            default:
                return 6;
        }
        
        // test with comments before every case
        var test7 = "d";
        switch (test7)
        {
            // comment before a
            case "a":
                return 1;
            // comment before b
            case "b":
                return 2;
            // comment before c
            case "c":
                return 3;
            /* covers case "d", "e": */
            // comment before default
            // comment before d
            // comment before e
            default:
                return 5;
            // comment before f
            case "f":
                return 6;
        }
    }
}
```

If I understood the current tests correctly, it's just about comparing the .as to .hx output. Is there any kind of test implemented currently where I could just run the `switch` statement with a number of values, and then expect a certain output?

### CODE REVIEW

File: src/as3hx/Writer.hx [hunk 1]
```diff
@@ -1306,6 +1306,14 @@ class Writer
                 
                 lvl++;
                 for(c in newCases) {
+
+                    if(def != null &&
+                        def.before != null &&
+                        def.before.el.toString() == c.el.toString()) {
```
Comment (mastef) [inline src/as3hx/Writer.hx]: I don't really like this `el.toString()` comparison, but it works. Any better ideas to compare `el`?

File: src/as3hx/Writer.hx [hunk 2]
```diff
@@ -1276,7 +1276,14 @@ class Writer
                         f(def.el[def.el.length - 1], def.el);
                     }
                 }
-                newCases = loopCases(cases.slice(0), def == null ? null : def.el.slice(0), testVar, newCases);
+
+                if (def != null && def.before == null) {
+                    // default is in the end
+                    newCases = loopCases(cases.slice(0), def.el.slice(0), testVar, newCases);
```
Comment (SlavaRa) [inline src/as3hx/Writer.hx]: Maybe you will replace `slice(0)` to `copy()`?

File: src/as3hx/parsers/StructureParser.hx [hunk 3]
```diff
@@ -185,14 +185,40 @@ class StructureParser {
...
+                            }
+                            // normal case, default not set yet, or differs
+                            else {
+                                var caseObj = { val : val, el : el, meta : meta }
+                                // default already set, but case follows it
+                                // mark that default is before this case
+                                if(def != null && def.before == null) {
+                                    def.before = caseObj;
+                                }
+                                cl.push( caseObj );
```
Comment (SlavaRa) [inline src/as3hx/parsers/StructureParser.hx]: please `cl.push( caseObj );` -> `cl.push(caseObj);`

**Comment by SlavaRa** [comment]
You should add test method to https://github.com/HaxeFoundation/as3hx/blob/master/test/unit/as3hx/AS3HXTest.hx

File: src/as3hx/Writer.hx (hunk 2 above)
Comment (mastef) [reply src/as3hx/Writer.hx]: Yeah I only adjusted the original code here - before it was `newCases = loopCases(cases.slice(0), def == null ? null : def.el.slice(0), testVar, newCases);
`
So it had before : `def == null ? null : def.el.slice(0)`

Now you'd like it to be `def.el.copy()[0]` ?

File: src/as3hx/parsers/StructureParser.hx (hunk 3 above)
Comment (mastef) [reply src/as3hx/parsers/StructureParser.hx]: The original had spaces as well 😅  `cl.push( { val : val, el : el, meta : meta } );`

I'll change it

**Comment by mastef** [comment]
Oh, I must have missed that in the previous PR as well!

File: src/as3hx/Writer.hx (hunk 2 above)
Comment (SlavaRa) [reply src/as3hx/Writer.hx]: Nope
`def.el.copy()` instead of `def.el.slice(0)`

File: src/as3hx/Writer.hx [hunk 4]
```diff
@@ -1276,7 +1276,14 @@ class Writer
...
                     }
                 }
-                newCases = loopCases(cases.slice(0), def == null ? null : def.el.slice(0), testVar, newCases);
+
+                if (def != null && def.before == null) {
+                    // default is in the end
+                    newCases = loopCases(cases.slice(0), def.el.copy(), testVar, newCases);
+                } else {
+                    // default is not in the end, so don't catch fall-through
+                    newCases = loopCases(cases.slice(0), null, testVar, newCases);
```
Comment (SlavaRa) [inline src/as3hx/Writer.hx]: `cases.slice(0)` -> `cases.copy()` ?

Comment (mastef) [reply src/as3hx/Writer.hx]: So just you're aware I'm not sure what I'm changing here. Can you propose a test-case?

**Comment by mastef** [comment]
Ok, all requested changes done

**Comment by SlavaRa** [comment]
Thx

**Pull request merged**